src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/CodeBasedDependencyAlgorithm.cs

```diff
@@ -118,23 +120,77 @@ public static void AddDependenciesDueToMethodCodePresence(ref DependencyList dep
             }
         }
 
-        private static void AddPInvokeParameterDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc parameter)
+        public static void AddInterestingInteropConstructedTypeDependencies(ref DependencyList dependencies, NodeFactory factory, TypeDesc type)
         {
-            if (parameter.IsDelegate)
+            if (type.IsDelegate)
             {
-                dependencies.Add(factory.NecessaryTypeSymbol(parameter), "Delegate Marshalling Stub");
+                var delegateType = type as MetadataType;
+                if (delegateType != null && delegateType.HasCustomAttribute("System.Runtime.InteropServices", "UnmanagedFunctionPointerAttribute"))
+                {
+                    AddDependenciesDuePInvokeDelegate(ref dependencies, factory, delegateType);
+                }
+            }
+            else if (type.IsValueType && type.IsTypeDefinition && !(type is NativeStructType))
+            {
+                var structType = type as MetadataType;
+                if (structType != null && structType.HasCustomAttribute("System.Runtime.InteropServices", "StructLayoutAttribute"))
```

Comment:
You are right. This path will never be taken. We need an attribute to indicate a struct is interesting for interop just likewe have UnmanagedFunctionPointer for delegates. I will comment this out and add a TODO to add an attribute for this.